apps/server/src/shared/repo/task/task.repo.ts

```diff
@@ -1,14 +1,77 @@
 import { Injectable } from '@nestjs/common';
-import { EntityManager, FilterQuery, QueryOrderMap } from '@mikro-orm/core';
+import { FilterQuery, QueryOrderMap } from '@mikro-orm/core';
+import { EntityManager } from '@mikro-orm/mongodb';
 
-import { EntityId, IFindOptions, Task, Counted } from '@shared/domain';
+import { EntityId, IFindOptions, Task, Counted, SortOrder } from '@shared/domain';
 
 import { TaskScope } from './task-scope';
 
 @Injectable()
 export class TaskRepo {
 	constructor(private readonly em: EntityManager) {}
 
+	async findAllFinishedByParentIds(
+		parentIds: {
+			creatorId: EntityId;
+			openCourseIds: EntityId[];
+			lessonIdsOfOpenCourses: EntityId[];
+			finishedCourseIds: EntityId[];
+			lessonIdsOfFinishedCourses: EntityId[];
+		},
+		options?: IFindOptions<Task>
+	): Promise<Counted<Task[]>> {
+		const { pagination } = options || {};
+
+		const scope = new TaskScope('$or');
+
+		const parentsOpen = new TaskScope('$or');
+		parentsOpen.byCourseIds(parentIds.openCourseIds);
+		parentsOpen.byLessonIds(parentIds.lessonIdsOfOpenCourses);
+
+		const parentsFinished = new TaskScope('$or');
+		parentsFinished.byCourseIds(parentIds.finishedCourseIds);
+		parentsFinished.byLessonIds(parentIds.lessonIdsOfFinishedCourses);
+
+		const openScope = new TaskScope();
+		openScope.addQuery(parentsOpen.query);
+		openScope.byDraft(false);
+		openScope.byClosed(parentIds.creatorId, true);
+
+		const finishedScope = new TaskScope();
+		finishedScope.addQuery(parentsFinished.query);
+		finishedScope.byDraft(false);
+
+		// must find also closed without course or lesson as parent
+		const creatorOpenScope = new TaskScope();
+		creatorOpenScope.byClosed(parentIds.creatorId, true);
+		creatorOpenScope.byCreatorId(parentIds.creatorId);
+
+		const creatorFinishedScope = new TaskScope();
+		creatorFinishedScope.addQuery(parentsFinished.query);
+		creatorFinishedScope.byCreatorId(parentIds.creatorId);
+
+		const creatorScope = new TaskScope('$or');
+		creatorScope.addQuery(creatorOpenScope.query);
+		creatorScope.addQuery(creatorFinishedScope.query);
+
+		scope.addQuery(openScope.query);
+		scope.addQuery(finishedScope.query);
+		scope.addQuery(creatorScope.query);
```

Comment:
maybe this would be more readable (and understandable):
```typescript
scope.addQuery(finishedByTeacherWithOpenCourseOrLessonScope.query);
scope.addQuery(finishedByCourseOrLesson.query);
scope.addQuery(finishedByTeacherOnlyScope.query);
```
however, we have to talk about the scopes and queries.

Reply:
You want to point out the renaming?

Reply:
I update it in a way that it discribe what it do. But the long names only describe what the by... scope names already discribe. With long names is come hard to readable and understand the primary points that we have 3 scopes and one of them is split internally to 2 scopes.